### PR TITLE
PREF-314 | add UseCase71 for proving deprecated tag

### PR DESCRIPTION
**Pull request by jpmarcotte**
Here's the relevant files it generates.
<details><summary>HasDeprecatedFields.php</summary>

```php
<?php
declare(strict_types=1);

namespace Neighborhoods\PrefabFitnessUseCase71\V1;

use JetBrains\PhpStorm\Deprecated;

class HasDeprecatedFields implements HasDeprecatedFieldsInterface
{
    /** @var int */
    private $id;

    /** @var bool */
    private $deprecated_field;

    /** @var string */
    private $replacement_field;

    /** @var bool */
    private $deprecated_simply;

    /** @var bool */
    private $deprecated_without_replacement;

    /** @var bool */
    private $deprecated_without_message;

    public function getId(): int
    {
        if ($this->id === null) {
            throw new \LogicException('id has not been set');
        }
        
        return $this->id;
    }
    
    public function setId(int $id): HasDeprecatedFieldsInterface
    {
        if ($this->id !== null) {
            throw new \LogicException('id has already been set');
        }
        
        $this->id = $id;
        return $this;
    }
    
    public function hasId(): bool
    {
        return $this->id !== null;
    }
     

    #[Deprecated(reason: 'in favor of replacement_field', replacement: '%class%->getReplacementField()')]
    public function getDeprecatedField(): bool
    {
        if ($this->deprecated_field === null) {
            throw new \LogicException('deprecated_field has not been set');
        }
        
        return $this->deprecated_field;
    }
    
    #[Deprecated(reason: 'in favor of replacement_field', replacement: '%class%->setReplacementField(%parameter0%)')]
    public function setDeprecatedField(bool $deprecated_field): HasDeprecatedFieldsInterface
    {
        if ($this->deprecated_field !== null) {
            throw new \LogicException('deprecated_field has already been set');
        }
        
        $this->deprecated_field = $deprecated_field;
        return $this;
    }
    
    #[Deprecated(reason: 'in favor of replacement_field', replacement: '%class%->hasReplacementField()')]
    public function hasDeprecatedField(): bool
    {
        return $this->deprecated_field !== null;
    }
     

    public function getReplacementField(): string
    {
        if ($this->replacement_field === null) {
            throw new \LogicException('replacement_field has not been set');
        }
        
        return $this->replacement_field;
    }
    
    public function setReplacementField(string $replacement_field): HasDeprecatedFieldsInterface
    {
        if ($this->replacement_field !== null) {
            throw new \LogicException('replacement_field has already been set');
        }
        
        $this->replacement_field = $replacement_field;
        return $this;
    }
    
    public function hasReplacementField(): bool
    {
        return $this->replacement_field !== null;
    }
     

    #[Deprecated()]
    public function getDeprecatedSimply(): bool
    {
        if ($this->deprecated_simply === null) {
            throw new \LogicException('deprecated_simply has not been set');
        }
        
        return $this->deprecated_simply;
    }
    
    #[Deprecated()]
    public function setDeprecatedSimply(bool $deprecated_simply): HasDeprecatedFieldsInterface
    {
        if ($this->deprecated_simply !== null) {
            throw new \LogicException('deprecated_simply has already been set');
        }
        
        $this->deprecated_simply = $deprecated_simply;
        return $this;
    }
    
    #[Deprecated()]
    public function hasDeprecatedSimply(): bool
    {
        return $this->deprecated_simply !== null;
    }
     

    #[Deprecated(reason: 'to prove this works')]
    public function getDeprecatedWithoutReplacement(): bool
    {
        if ($this->deprecated_without_replacement === null) {
            throw new \LogicException('deprecated_without_replacement has not been set');
        }
        
        return $this->deprecated_without_replacement;
    }
    
    #[Deprecated(reason: 'to prove this works')]
    public function setDeprecatedWithoutReplacement(bool $deprecated_without_replacement): HasDeprecatedFieldsInterface
    {
        if ($this->deprecated_without_replacement !== null) {
            throw new \LogicException('deprecated_without_replacement has already been set');
        }
        
        $this->deprecated_without_replacement = $deprecated_without_replacement;
        return $this;
    }
    
    #[Deprecated(reason: 'to prove this works')]
    public function hasDeprecatedWithoutReplacement(): bool
    {
        return $this->deprecated_without_replacement !== null;
    }
     

    #[Deprecated(replacement: '%class%->getId()')]
    public function getDeprecatedWithoutMessage(): bool
    {
        if ($this->deprecated_without_message === null) {
            throw new \LogicException('deprecated_without_message has not been set');
        }
        
        return $this->deprecated_without_message;
    }
    
    #[Deprecated(replacement: '%class%->setId(%parameter0%)')]
    public function setDeprecatedWithoutMessage(bool $deprecated_without_message): HasDeprecatedFieldsInterface
    {
        if ($this->deprecated_without_message !== null) {
            throw new \LogicException('deprecated_without_message has already been set');
        }
        
        $this->deprecated_without_message = $deprecated_without_message;
        return $this;
    }
    
    #[Deprecated(replacement: '%class%->hasId()')]
    public function hasDeprecatedWithoutMessage(): bool
    {
        return $this->deprecated_without_message !== null;
    }
     

    public function jsonSerialize()
    {
        return get_object_vars($this);
    }
}
```
</details>

<details><summary>HasDeprecatedFieldsInterface.php</summary>

```php
<?php
declare(strict_types=1);

namespace Neighborhoods\PrefabFitnessUseCase71\V1;

use JetBrains\PhpStorm\Deprecated;

interface HasDeprecatedFieldsInterface extends \JsonSerializable
{
    public const TABLE_NAME = 'unused';

    public const PROP_ID = 'id';
    #[Deprecated(reason: 'in favor of replacement_field', replacement: '%class%::PROP_REPLACEMENT_FIELD')]
    public const PROP_DEPRECATED_FIELD = 'deprecated_field';
    public const PROP_REPLACEMENT_FIELD = 'deprecated_field';
    #[Deprecated()]
    public const PROP_DEPRECATED_SIMPLY = 'deprecated_simply';
    #[Deprecated(reason: 'to prove this works')]
    public const PROP_DEPRECATED_WITHOUT_REPLACEMENT = 'deprecated_without_replacement';
    #[Deprecated(replacement: '%class%::PROP_ID')]
    public const PROP_DEPRECATED_WITHOUT_MESSAGE = 'deprecated_without_reason';

    public function getId(): int;
    public function setId(int $id): HasDeprecatedFieldsInterface;
    public function hasId(): bool;

    #[Deprecated(reason: 'in favor of replacement_field', replacement: '%class%->getReplacementField()')]
    public function getDeprecatedField(): bool;
    #[Deprecated(reason: 'in favor of replacement_field', replacement: '%class%->setReplacementField(%parameter0%)')]
    public function setDeprecatedField(bool $deprecated_field): HasDeprecatedFieldsInterface;
    #[Deprecated(reason: 'in favor of replacement_field', replacement: '%class%->hasReplacementField()')]
    public function hasDeprecatedField(): bool;

    public function getReplacementField(): string;
    public function setReplacementField(string $replacement_field): HasDeprecatedFieldsInterface;
    public function hasReplacementField(): bool;

    #[Deprecated()]
    public function getDeprecatedSimply(): bool;
    #[Deprecated()]
    public function setDeprecatedSimply(bool $deprecated_simply): HasDeprecatedFieldsInterface;
    #[Deprecated()]
    public function hasDeprecatedSimply(): bool;

    #[Deprecated(reason: 'to prove this works')]
    public function getDeprecatedWithoutReplacement(): bool;
    #[Deprecated(reason: 'to prove this works')]
    public function setDeprecatedWithoutReplacement(bool $deprecated_without_replacement): HasDeprecatedFieldsInterface;
    #[Deprecated(reason: 'to prove this works')]
    public function hasDeprecatedWithoutReplacement(): bool;

    #[Deprecated(replacement: '%class%->getId()')]
    public function getDeprecatedWithoutMessage(): bool;
    #[Deprecated(replacement: '%class%->setId(%parameter0%)')]
    public function setDeprecatedWithoutMessage(bool $deprecated_without_message): HasDeprecatedFieldsInterface;
    #[Deprecated(replacement: '%class%->hasId()')]
    public function hasDeprecatedWithoutMessage(): bool;
}
```
</details>

<details><summary>Screenshots of Builder.php</summary>

Hovering over `setDeprecatedField`:
image

Opt+Enter on `setDeprecatedField`:
image

</details>

### CODE REVIEW

File: UseCase71/composer.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    "prefab"
+  ],
+  "config": {
+    "sort-packages": true,
+    "platform": {
+      "php": "7.3"
+    }
+  },
+  "require": {
+    "neighborhoods/prefab": "dev-PREF-314-7x-deprecated-flag"
```
Comment (jpmarcotte) [inline UseCase71/composer.json:15]: TODO: Update this after the Prefab PR is merged.

**Review by arielallon — APPROVED** [review]

**Review by mucha55 — APPROVED** [review]
I think this should contain another example field that is deprecated, but without a deprecation message (to demonstrate that it's optional). Not a hard blocker

**Review by mucha55 — APPROVED** [review]
Same non-blocking comment as before (only bringing it up because you 👍'd it previously):

> I think this should contain another example field that is deprecated, but without a deprecation message (to demonstrate that it's optional)

File: UseCase71/src/V1/HasDeprecatedFields.prefab.definition.yml [hunk 2]
```diff
@@ -11,10 +11,27 @@ properties:
...
+    is_deprecated: true
+  deprecated_without_replacement:
+    data_type: bool
+    record_key: deprecated_without_replacement
+    nullable: true
+    is_deprecated: true
+    deprecated_message: 'to prove this works'
+  deprecated_without_message:
+    data_type: bool
+    record_key: deprecated_without_reason
```
Comment (mucha55) [inline UseCase71/src/V1/HasDeprecatedFields.prefab.definition.yml:34]: Possibly intentional but the `record_key` doesn't match the field name

**Review by mucha55 — APPROVED** [review]
💪

File: UseCase71/src/V1/HasDeprecatedFields.prefab.definition.yml (hunk 2 above)
Comment (jpmarcotte) [reply UseCase71/src/V1/HasDeprecatedFields.prefab.definition.yml:34]: Heh, not intentional. I changed one and forgot to change the other.

**Review by arielallon — APPROVED** [review]

**Review by arielallon — APPROVED** [review]

**Review by mucha55 — APPROVED** [review]

**Review by arielallon — APPROVED** [review]

**Comment by arielallon** [comment]
actually, if this is intended as a test suite, do we want the version match to be a little looser, like `^7.2.0` to ensure that nothing in the major version line breaks it?
(and updated to `^7.2.0|^8.x.y` when this makes its way into the 8.x line)

**Comment by jpmarcotte** [comment]
I don't think this should really be intended as a test suite so much as a "how to do things" (primary) and a rough proving grounds for developing new features (secondary).

I do think that Prefab needs something along the lines of what Buphalo has for full-blown integration tests, but that's probably a different task.

**Comment by jpmarcotte** [comment]
> (and updated to `^7.2.0 || ^8.x.y` when this makes its way into the 8.x line)

For that, I'll probably create a new `UseCase8X`.

**Pull request merged**